### PR TITLE
Support dynamic provisioning for CSI migration scenarios

### PR DESCRIPTION
**Pull request by ddebroy**
Part of enhancements for [kubernetes/enhancements#625](https://github.com/kubernetes/enhancements/issues/625)

This PR depends on https://github.com/kubernetes/kubernetes/pull/73653 for a set of changes needed in in-tree PV Controller to call out to the external provisioning controller for in-tree plugins if they are migratable to CSI. The basic steps in that PR are:

1. Check if in-tree plugin is migratable to CSI. If so:
2. Set `annStorageProvisioner` in the PVC to CSI plugin name that superseded the specified in-tree plugin.
3. Do not proceed with in-tree provisioning and return so that external provisioner can act on the claim.

This PR introduces the following changes in lib-external-provisioner in the context of dynamic provisioning in CSI migration scenarios to act on the above annotated claims:

1. Enhances `getStorageClassFields` to check (through `GetCSINameFromIntreeName`) if the storageclass provisioner has a corresponding CSI plugin that supersedes it. If so:
2. Makes `getStorageClassFields` aware that it's handling dynamic volume provisioning for an in-tree plugin that has been migrated to CSI because [a] For regular in-tree provisioning path, external provisioning will never get invoked [b] For regular CSI provisioning path, `GetCSINameFromIntreeName` would not return anything for a CSI plugin name.
3. Enhances `getStorageClassFields` to [a] translate in-tree storage class parameters to the corresponding CSI plugin parameters using `TranslateInTreeStorageClassParametersToCSI` [b] return the CSI plugin name (instead of the storage class in-tree provisioner) as the provisioner name (so that it aligns with `ctrl.provisionerName`) and [c] return a boolean to `provisionClaimOperation` indicating dynamic provisioning in a CSI migration context
4. Enhances `provisionClaimOperation` to invoke `TranslateCSIPVToInTree` for dynamic provisioning in a CSI migration context.
5. A Unit test to ensure a storage class with in-tree provisioner can delegate to external provisioner for provisioning and translate the PV back to in-tree source.

/sig storage
/assign @jsafrane @saad-ali
/cc @msau42 @leakingtapan @davidz627

### CODE REVIEW

**Comment by jsafrane** [comment]
I'd prefer if this translation is done in kubernetes-csi/external-provisioner instead of this library. Is there any good reason to put it here?

**Comment by ddebroy** [comment]
@jsafrane The changes in `getStorageClassFields()` are required here in the library because of the check at https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go#L1039 which will reject the claim if storage class provisioner and provisioner name does not match up. For CSI migration scenarios, they will not (since storage class provisioner will point to in-tree while ext. provisioner will point to CSI). As a result, the provision() call to the CSI external provisioner will not come through.

The calls to `TranslateInTreeStorageClassParametersToCSI` and `TranslateCSIPVToInTree` are pretty tightly coupled with the above change/logic in `getStorageClassFields()` which is why I also placed them in the library. They may be decoupled to the CSI external provisioner if desired but may be hard to follow later on.

**Comment by jsafrane** [comment]
I think that the provisioner library should be as generic as possible and should not know anything about CSI. You're right, `provisionerName` check stands in the way.

Would it be cleaner to have an array of provisioner names instead of one in the provisioner lib + the rest in CSI external-provisioner?

@wongma7 @msau42 @davidz627

**Comment by msau42** [comment]
Configuring the lib with the provisioner names to respond to sounds reasonable to me
